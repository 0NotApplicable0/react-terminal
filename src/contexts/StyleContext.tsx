import * as React from "react"
import * as style from "../index.scss"


export const StyleContext = React.createContext(null)


export const StyleContextProvider = (props: any) => {

    return (
        <StyleContext.Provider value={style.default}>
            {props.children}
        </StyleContext.Provider>
    )
}

export default {
    StyleContext,
    StyleContextProvider
}