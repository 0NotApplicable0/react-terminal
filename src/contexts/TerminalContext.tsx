import * as React from "react";

export const TerminalContext = React.createContext(null);

export const TerminalContextProvider = (props: any) => {
  const { children } = props;
  const [bufferedContent, setBufferedContent] = React.useState("");
  const [commandsHistory, setCommandsHistory] = React.useState([]);
  const [historyPointer, setHistoryPointer] = React.useState(-1);

  const appendCommandToHistory = (command: string) => {
    if (!command) {
      return;
    }

    setHistoryPointer(commandsHistory.length);
    setCommandsHistory(commandsHistory.concat(command));
  };

  const getPreviousCommand = () => {
    if (historyPointer === -1) {
      return "";
    }

    const command = commandsHistory[historyPointer];
    if (historyPointer > 0) {
      setHistoryPointer(historyPointer - 1);
    }

    return command;
  };

  const getNextCommand = () => {
    if (historyPointer + 1 === commandsHistory.length) {
      return "";
    }

    const command = commandsHistory[historyPointer + 1];
    setHistoryPointer(historyPointer + 1);
    return command;
  };

  return (
    <TerminalContext.Provider
      value={{
        bufferedContent,
        setBufferedContent,
        appendCommandToHistory,
        getPreviousCommand,
        getNextCommand
      }}
    >
      {children}
    </TerminalContext.Provider>
  );
};

export default {
  TerminalContext,
  TerminalContextProvider
};
