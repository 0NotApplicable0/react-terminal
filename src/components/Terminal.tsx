import * as React from "react";
import { StyleContext } from "../contexts/StyleContext";
import { useClickOutsideEvent } from "../hooks/terminal";

import Controls from "./Controls";
import Editor from "./Editor";

export default function Terminal(props: any) {
  const wrapperRef = React.useRef(null);
  const [consoleFocused, setConsoleFocused] = React.useState(true);
  const style = React.useContext(StyleContext);

  useClickOutsideEvent(wrapperRef, setConsoleFocused);

  // Get all props destructively
  const {
    theme,
    showControlButtons,
    controlButtonLabels,
    prompt,
    commands,
    errorMessage
  } = props;

  return (
    <div
      ref={wrapperRef}
      id={style.terminalContainer}
      className={style[`theme--${theme}`]}
    >
      <div className={`${style.terminal}`}>
        <Controls
          consoleFocused={consoleFocused}
          showControlButtons={showControlButtons}
          controlButtonLabels={controlButtonLabels}
        />
        <Editor
          consoleFocused={consoleFocused}
          prompt={prompt}
          commands={commands}
          errorMessage={errorMessage}
        />
      </div>
    </div>
  );
}

Terminal.defaultProps = {
  theme: "light",
  showControlButtons: true,
  controlButtonLabels: ["close", "minimize", "maximize"],
  prompt: ">>>",
  commands: {},
  errorMessage: "not found!",
};
