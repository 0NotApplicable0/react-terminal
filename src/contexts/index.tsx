import * as React from "react";
import { StyleContextProvider } from "./StyleContext";

export default function ContextProvider(props: any) {
  const { children } = props;

  return <StyleContextProvider>{children}</StyleContextProvider>;
}
