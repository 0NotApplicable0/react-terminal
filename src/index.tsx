import * as React from 'react';
import * as Terminal from './components/Terminal';
import * as ContextProvider from './contexts';


export default function ReactTerminal(props: any): any {
  return (
    <ContextProvider.default>
      <Terminal.default props={props} />
    </ContextProvider.default>
  );
}
