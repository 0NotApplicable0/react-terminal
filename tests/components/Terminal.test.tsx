import * as React from "react";
import renderer from "react-test-renderer";
import * as ContextProvider from "../../src/contexts";
import Terminal from "../../src/components/Terminal";

let props: any;
const renderWrapper = () =>
  renderer.create(
    <ContextProvider.default>
      <Terminal {...props} />
    </ContextProvider.default>
  );

it("Terminal renders correctly", () => {
  expect(renderWrapper().toJSON()).toMatchSnapshot();
});
