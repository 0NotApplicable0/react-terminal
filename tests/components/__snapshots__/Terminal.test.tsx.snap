// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal renders correctly 1`] = `
<div
  className="theme--light"
  id="terminalContainer"
>
  <div
    className="terminal"
  >
    <div
      className="controls"
    >
      <div
        className="consoleCtrl close"
      />
      <div
        className="consoleCtrl minimize"
      />
      <div
        className="consoleCtrl maximize"
      />
    </div>
    <div
      className="editor"
    >
      
      <span
        className="prompt"
      >
        &gt;&gt;&gt;
      </span>
      <div
        className="lineText"
      >
        <span
          className="preWhiteSpace"
        >
          
        </span>
        <span
          className="caret"
        >
          <span
            className="caretAfter"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
